bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/JSONResponse.java

```diff
@@ -64,79 +103,75 @@ private static ResponseBuilder response(Response.Status status) {
         }
 
-        return ret;
+        return resultJson;
     }
 
-    /**
-     * in case of error (404 and such)
-     *
-     * @param status
-     * @param errormessage
-     * @return Response containing a status and the errormessage in JSON format
-     */
-    public static Response createErrorResponse(Response.Status status, String errormessage) {
-        return createResponse(status, null, errormessage);
+    private Response createErrorResponse(Response.Status status, Object entity, String errormessage) {
+        ResponseBuilder rp = responseBuilder(status);
+        JsonElement errorJson = createErrorJson(errormessage, status, entity, null);
+        rp.entity(errorJson);
+        return rp.build();
     }
 
-    /**
-     * Depending in the status, create a Response object containing either the entity alone or an error JSON
-     * which might hold the entity as well.
-     *
-     * @param status
-     * @param entity
-     * @param errormessage an optional error message (may be null), ignored if the status family is successful
-     * @return Response configure for error or success
-     */
-    public static Response createResponse(Response.Status status, Object entity, String errormessage) {
-        JsonElement ret;
-        if (status.getFamily() == Response.Status.Family.SUCCESSFUL) {
-            // create non-null JsonElement if null!=entity
-            ret = null != entity ? GSON.toJsonTree(entity) : null;
-        } else {
-            ret = createErrorJson(errormessage, status, entity, null);
+    private Response createResponse(Status status, Object entity) {
+        ResponseBuilder rp = responseBuilder(status);
+        PipedOutputStream out = new PipedOutputStream();
+
+        try {
+            // we will not actively close the PipedInputStream since it is read by the receiving end
+            // and will be GC'ed once the response is consumed.
+            PipedInputStream in = new PipedInputStream(out);
+            rp.entity(in);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
         }
 
-        // configure response
-        ResponseBuilder rp = response(status);
-        if (null != ret) {
-            rp = rp.entity(GSON.toJson(ret));
-        }
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                OutputStreamWriter outStreamWriter = new OutputStreamWriter(out);
+                try (JsonWriter jsonWriter = GSON.newJsonWriter(new BufferedWriter(outStreamWriter))) {
+                    if (entity != null) {
+                        GSON.toJson(entity, entity.getClass(), jsonWriter);
+                        jsonWriter.flush();
+                    }
+                } catch (IOException | JsonIOException e) {
+                    logger.error("Error streaming JSON through PipedInpuStream/PipedOutputStream: ", e);
+                } finally {
+                    IOUtils.closeQuietly(outStreamWriter);
```

Comment:
Any reason not to use try-with-resources for `outStreamWriter`?
You don't need another try-block but can add it to the current one.

```Java
            public void run() {
                try (final OutputStreamWriter outStreamWriter = new OutputStreamWriter(out);
                        final JsonWriter jsonWriter = GSON.newJsonWriter(new BufferedWriter(outStreamWriter))) {
                    if (entity != null) {
                        GSON.toJson(entity, entity.getClass(), jsonWriter);
                        jsonWriter.flush();
                    }
                } catch (IOException | JsonIOException e) {
                    logger.error("Error streaming JSON through PipedInpuStream/PipedOutputStream: ", e);
                }
            }
```